pkg/codegen/hcl2/model/expression.go

```diff
@@ -1050,17 +1050,30 @@ func (x *FunctionCallExpression) Typecheck(typecheckOperands bool) hcl.Diagnosti
 	typecheckDiags := typecheckArgs(rng, x.Signature, x.Args...)
 	diagnostics = append(diagnostics, typecheckDiags...)
 
-	// TODO(https://github.com/pulumi/pulumi/issues/8439): This lifting is overly aggressive, we special case this for
-	// pulumiResourceName and pulumiResourceType as they should return a plain string even with an output input. But
-	// this should be covered more generally, only lifting when needed.
-	//
-	// Unless the function is already automatically using an Output-returning version, modify the signature to account
-	// for automatic lifting to Promise or Output.
-	if x.Name != "pulumiResourceName" && x.Name != "pulumiResourceType" {
-		_, isOutput := x.Signature.ReturnType.(*OutputType)
-		if !isOutput {
-			x.Signature.ReturnType = liftOperationType(x.Signature.ReturnType, x.Args...)
+	// If any of the inputs are Output<T> but the function only expects T then we need to lift the function into output
+	// space.
+	lift := false
+	for i, arg := range x.Args {
+		var param Parameter
+		if i >= len(x.Signature.Parameters) {
+			if x.Signature.VarargsParameter == nil {
+				// If we get here we must have already generated an error diagnostic above, just break out the loop at
+				// this point.
+				break
+			}
+			param = *x.Signature.VarargsParameter
+		} else {
+			param = x.Signature.Parameters[i]
 		}
+
+		if !param.Type.AssignableFrom(arg.Type()) {
```

Comment:
I'm guessing that the only way they are not assignable is T vs Output<T>? How do we know this to be the case -- has something prior to this validated this?

Reply:
Yeh we'll have validated when binding the function, so the only differences that we'll see here are output vs plain.